From 03b4febfd10a1a0d118f91bc9c7ecab6ac7db792 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Fri, 14 Jun 2019 12:59:12 +0700
Subject: [PATCH] drm: rcar-du: Add HDMI resolution check for RZ/G2

Because the maximum resolution of HDMI built in RZ/G2 is
3840x2160@30Hz (dotclock is up to 297MHz), the resolution exceeding
that is to return MODE_BAD.

This commit based on commit applied for R-Car Gen3:
https://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git/commit/drivers/gpu/drm/rcar-du/rcar_dw_hdmi.c?id=628daa03f3176714295209c8637e96a5ae08c4da

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_dw_hdmi.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/gpu/drm/rcar-du/rcar_dw_hdmi.c b/drivers/gpu/drm/rcar-du/rcar_dw_hdmi.c
index 76210ae..318a7b9 100644
--- a/drivers/gpu/drm/rcar-du/rcar_dw_hdmi.c
+++ b/drivers/gpu/drm/rcar-du/rcar_dw_hdmi.c
@@ -16,6 +16,8 @@
 
 #include <drm/bridge/dw_hdmi.h>
 
+#include "rcar_du_drv.h"
+
 #define RCAR_HDMI_PHY_OPMODE_PLLCFG	0x06	/* Mode of operation and PLL dividers */
 #define RCAR_HDMI_PHY_PLLCURRGMPCTRL	0x10	/* PLL current and Gmp (conductance) */
 #define RCAR_HDMI_PHY_PLLDIVCTRL	0x11	/* PLL dividers */
@@ -62,8 +64,32 @@ static int rcar_hdmi_phy_configure(struct dw_hdmi *hdmi,
 	return 0;
 }
 
+static enum drm_mode_status rcar_hdmi_mode_valid(
+					struct drm_connector *connector,
+					const struct drm_display_mode *mode)
+{
+	struct drm_device *ddev = connector->dev;
+	struct rcar_du_device *rcdu = ddev->dev_private;
+
+	if (rcdu->info->gen != 3)
+		return MODE_OK;
+
+	if (mode->hdisplay > 3840 || mode->vdisplay > 2160)
+		return MODE_BAD;
+
+	if (mode->hdisplay == 3840 && mode->vdisplay == 2160 &&
+	    mode->vrefresh > 30)
+		return MODE_BAD;
+
+	if (mode->clock > 297000)
+		return MODE_BAD;
+
+	return MODE_OK;
+}
+
 static const struct dw_hdmi_plat_data rcar_dw_hdmi_plat_data = {
 	.configure_phy	= rcar_hdmi_phy_configure,
+	.mode_valid = rcar_hdmi_mode_valid,
 };
 
 static int rcar_dw_hdmi_probe(struct platform_device *pdev)
-- 
2.7.4

